src/core/lib/passport.ts

```diff
@@ -52,13 +68,35 @@ passport.use(
             });
           }
 
+          // Check if password needs to be rehashed
           if (contact.password.iterations < config.passwordIterations) {
             await ContactsService.updateContact(contact, {
               password: await generatePassword(password)
             });
           }
 
-          return done(null, contact, { message: "logged-in" });
+          if (contact.otp.key) {
+            log.warn("The user has the old 2FA enabled.");
+          }
+
+          const mfa = await ContactMfaService.get(contact);
+
+          // Check if multi factor authentication is enabled and supported
+          if (mfa) {
+            if (mfa.type !== ContactMfaType.TOTP) {
+              log.warn("The user has unsupported 2FA enabled.");
+              // We pass the contact to the done callback so the user can be logged in and the 2FA is ignored
+              return done(null, contact, {
+                message: LOGIN_CODES.UNSUPPORTED_2FA
+              });
+            }
+
+            // Please see AuthController for the rest of this authentication flow
+            return done(null, contact, { message: LOGIN_CODES.REQUIRES_2FA });
```

Comment:
I think you should check the 2FA token in here. If you pass `contact` to `done` then it will create a session for the user as passport treats `done(null, contact)` as a success. Even if they can't use this session (could they?) I think it's safer to not create it at all.

Reply:
We could also pass an error (first parameter) here so that the user is not added to the session. Unfortunately, I can't get to the token at this point, so I would have to implement a separate strategy for MFA or adapt `passport-local`.

Reply:
Ahh I see. I had a quick look at the `passport-local` source code and it looks like there is a hidden option `passReqToCallback` that you could use to get access to the underlying request opject.

https://github.com/jaredhanson/passport-local/blob/master/lib/strategy.js#L113

Reply:
Thanks for finding 👍